### ISSUE TITLE
Should vector skins/costumes always be rendered with a resolution greater than their screen-space dimensions?

### ISSUE THREAD

**Issue by adroitwhiz** [LLK/scratch-render#620]
### Expected Behavior

Vector skins/costumes should appear sharp at all scales.

### Actual Behavior

Vector skins are internally represented by a series of textures at various resolutions. These textures are allowed to appear on-screen at up to 1.5x their resolution, meaning that at worst, 1 "texture pixel" will be magnified to 1.5 "screen pixels". This results in vector sprites often appearing blurry:
![image](https://user-images.githubusercontent.com/25993062/82130649-43b6cd00-979b-11ea-97b8-f8d9a9b20ba4.png)

This blurriness is particularly bad for things like text, which results in issues like https://github.com/LLK/scratch-render/issues/558.

This could be changed so that the textures must always be at a higher resolution than, or the same resolution as, their "screen-space" dimensions. This would consume more RAM, but the improvement in visual fidelity may well be worth it.

Keep in mind that until last October, [vector costumes were rendered at least twice as large as they should have been on high-DPI devices like phones](https://github.com/LLK/scratch-svg-renderer/pull/84), so this shouldn't have an unprecedented RAM impact.

### Steps to Reproduce

1. Create a new project
2. With the default cat sprite selected, click on the "change size by 10" block in the Blocks Palette while observing the cat sprite
3. Notice that as you continue clicking the block, the cat gets blurrier as it grows, eventually becoming crisp again if you click the block enough

### Operating System and Browser

All

**Comment by kchadha** [comment]
@cwillisf @fsih @adroitwhiz, assigned this to you for discussion (and then documenting here)

**Comment by adroitwhiz** [comment]
The fix itself is quite easy to implement-- I already have it up in a branch:
https://adroitwhiz.github.io/scratch-gui/no-vector-blur/

The hard part is benchmarking across devices to make sure the increased RAM usage is acceptable (and figuring out what "acceptable" means).

**Comment by tutacat** [comment]
@adroitwhiz Have you also looked at why the vector sprites have odd cropping?
(e.g. a single straight won't display, and a single line sometimes gets chopped somewhere)

**Comment by apple502j** [comment]
@tutacat https://github.com/LLK/scratch-svg-renderer/pull/85#issuecomment-639100645

**Comment by adroitwhiz** [comment]
@tutacat If you're talking about https://github.com/LLK/scratch-render/issues/639 or https://github.com/LLK/scratch-render/issues/427, that'll be fixed by https://github.com/LLK/scratch-render/pull/594

**Comment by fsih** [comment]
Discussion:
- It's very difficult to instrument the VRAM usage of graphics cards, so we probably wouldn't be able to do that
- We can use the built-in profiler on browsers to look at the total RAM usage and compare before and after
- We have a maximum texture size in the renderer that we wouldn't go above, so this wouldn't affect very large textures, only the average size of textures
- We recommend testing a relatively complex project with a large number of vector sprites and comparing RAM usage before and after the change on a very low-end device (probably the raspberry pi), and seeing if the change causes performance (frame rate) to become unacceptable. 
- We think it's worth putting in if the performance is acceptable

**Comment by fsih** [comment]
Next steps

@adroitwhiz do you have the PR for https://adroitwhiz.github.io/scratch-gui/no-vector-blur/ up somewhere for us to look at?

@BryceLTaylor could you run a project with a lot of vector sprites on a Raspberry Pi and compare if it's noticeably worse on https://adroitwhiz.github.io/scratch-gui/no-vector-blur/ vs prod?

**Comment by adroitwhiz** [comment]
Just created a PR for this change: https://github.com/LLK/scratch-render/pull/649

**Comment by BryceLTaylor** [comment]
I tested the branch on a Raspberry Pi 4 by playing a MAP (https://scratch.mit.edu/projects/398958757/).  On production after 20 or so seconds of playing the project the whole computer became unresponsive.  The mouse cursor moved several seconds after I had tried moving it, and then eventually stopped moving altogether.  I had to restart the whole machine. 
On the branch it played fine for ~2 minutes, but eventually the tab crashed.  

Also, on a project with just the default cat I saw some somewhat strange behavior when sprites were at 117% and 118% size, but I have been unable to reproduce it.  The cat does look a little less blurry to me than with a similar project on production.
